README.md
# Generating data for "Characterizing narrative time in books through fluctuations in power and danger arcs"

### Installing a conda environment

Clone this repo into your machine. Open a terminal in the folder containing all the downloaded repo files, and create a conda environment (which we name `ousiometrics`) using

```
conda create --name ousiometrics --file requirements.txt
conda activate ousiometrics
```

### Downloading text from Gutenberg
Download a text file from Project Gutenberg using the instructions from their website: https://www.gutenberg.org/help/mirroring.html.

As an example, we placed a downloaded file from Project Gutenberg in the `./demo/gutenberg_txt/` folder.

The Project Gutenberg IDs used in the study are given in `PG_IDs.csv`.

### Removing headers

We want to remove the headers of the files from Gutenberg. Assuming the downloaded text files from Gutenberg (with extensions beginning with `*.txt`) are in `$SRCDIR_ORIG` and you want to place the processed files in `$SRCDIR_CLEAN`, run

`python cleanup_gutenberg_headers.py $SRCDIR_ORIG $SRCDIR_CLEAN`

For the demonstration, we use `./demo/gutenberg_txt` as `$SRCDIR_ORIG` and `./demo/gutenberg_txt_clean` as `$SRCDIR_CLEAN`. The file `cleanup_gutenberg_headers.py` was adapted from the code in [[1]](#1).

### Generating the time series

The time series can be generated for a given window size $N_w$ and skip size $N_s$. In our analysis, we used $N_w=50$ and $N_s=50$. If we want to put the resulting files in `./demo/outputdir/`, the corresponding scores for every window can be generated by

```
python generate_texttrends_files.py ./demo/gutenberg_txt_clean/PG1342_text.txt ./demo/outputdir --N_w 50 --N_s 50 --overwrite
```

The `--overwrite` argument is included to overwrite the relevant output files in `./demo/*/`.

To generate the scores for shuffled text, one can use the additional arguments `--shuffle` and `--seed`. In this example, we set the seed to be `42`:

```
python generate_texttrends_files.py ./demo/gutenberg_txt_clean/PG1342_text.txt ./demo/outputdir --N_w 50 --N_s 50 --shuffle --seed 42 --overwrite
```

Note that new subfolders inside `./outputdir` are generated automatically.

### Generating the IMFs

To compute for the IMFs, the `emd` module [[2]](#1) has to be installed from the folder in this repo. This is because the `ensemble_sift` function was modified to allow for a seed as a keyword argument.

To install the modified `emd` module:

```
cd ./emd
pip install -r requirements.txt
pip install .
```

Return to the folder containing the file `get_hht_freqs.py`. To obtain the IMF files for the time series corresponding to a PG ID in some `$TS_FOLDER` (e.g., `1342` in folder `./demo/outputdir/window=50_n=None_skip=50_thresh=0.7_shuffle=False/`), run
```
python get_hht_freqs.py ./demo/outputdir/window=50_n=None_skip=50_thresh=0.7_shuffle=False/ 1342 --overwrite
```
The output subdirectories `hht/` and `imf/` containing the HHT and IMF-related files will be in `$TS_FOLDER`. Columns ending with a `_<number>` correspond to a given IMF order, with the last IMF computed by `emd` corresponding to the trend. Note that in Python, the first IMF is labeled `0` (i.e., `imf_0`, etc).

**References**

<a id="1">[1]</a> 
Gerlach, M., & Font-Clos, F. (2020). A standardized Project Gutenberg corpus for statistical analysis of natural language and quantitative linguistics. Entropy, 22(1), 126. (Code in https://github.com/pgcorpus/gutenberg)

<a id="2">[2]</a> 
https://emd.readthedocs.io/en/v0.4.0/

emd/changelog.md
# Changelog

Changes and updates to EMD are tracked by version on this page.  The format of
this changelog is (mostly) based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this the EMD package uses [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

Changes should be categorised under the following types:

- **Added** for new features.
- **Changed** for changes in existing functionality.
- **Deprecated** for soon-to-be removed features.
- **Removed** for now removed features.
- **Fixed** for any bug fixes.
- **Security** in case of vulnerabilities.

Where appropriate, links to specific Issues & Merge Requests on [our gitlab page](https://gitlab.com/emd-dev/emd).


## Development Version
Work in progress...

    git clone https://gitlab.com/emd-dev/emd.git

### Added
- Support for Python 3.9 - requirements updated and test build added [!62](https://gitlab.com/emd-dev/emd/-/merge_requests/62)
- New citations page on website (and assorted website/reference fixups) [!60](https://gitlab.com/emd-dev/emd/-/merge_requests/60)

### Fixed
- Average over only the ensembles with the modal number of IMFs in ensemble_sift, previously there could be an error due to difference in nimfs between ensembles [!63](https://gitlab.com/emd-dev/emd/-/merge_requests/63)
- Fix pyyaml dependency for google-colab import [!64](https://gitlab.com/emd-dev/emd/-/merge_requests/64)
- Fix case where mask sift could drop some options [!66](https://gitlab.com/emd-dev/emd/-/merge_requests/66)

---

# Stable Versions

## 0.4.0

    pip install emd==0.4.0
Released 2021-03-30

### Notes
Many changes in this revision come from the review process at [JOSS](https://github.com/openjournals/joss-reviews/issues/2977)

### Added
- New tutorials
  - Cross-frequency coupling [!52](https://gitlab.com/emd-dev/emd/-/merge_requests/52)
  - Why use EMD? [90a5b5e2](https://gitlab.com/emd-dev/emd/-/commit/90a5b5e2e4ffdd7634cf63e30836843c920fcaa3)
  - Tutorial on code speed [35ae8c82](https://gitlab.com/emd-dev/emd/-/commit/35ae8c82ab72b9c36d641eb4bef4d1ae7c53b0a5)
- Second layer mask sift function [65a05dd2](https://gitlab.com/emd-dev/emd/-/commit/65a05dd2cf1610508d13a68f6753094f07d67e48)
- Add html printing functionality for SiftConfig [5c57781e](https://gitlab.com/emd-dev/emd/-/commit/5c57781e2e8b92b8d2a7e00ceec8bde064bc412b)
- Update contribution and installation details on website - add accordions for better readability [!53](https://gitlab.com/emd-dev/emd/-/merge_requests/53)
- Add new plotting functionality for HHT and Holospectra [!53](https://gitlab.com/emd-dev/emd/-/merge_requests/53)
- Show warning when max_imfs is very high compared to length of time-series [4cd15291](https://gitlab.com/emd-dev/emd/-/commit/4cd15291c25e082cbb9ffb56a2c3812b6b3d391e)

### Changed
- Major refactor in handling of cycles analysis [!56](https://gitlab.com/emd-dev/emd/-/merge_requests/56)
  - Introduce Cycles class
  - Introduce \_cycles\_support module
- Renamed 'References' webpage to 'API' [d8fe93b5](https://gitlab.com/emd-dev/emd/-/commit/d8fe93b520c19ce45f3f5a73294074a4b1d75ce5)

### Fixed
- Widespread fixing of typos and mistakes in documentation & website [!52](https://gitlab.com/emd-dev/emd/-/merge_requests/52)
- Make docstrings pydocstyle compliant and add pydocstyle conventions [!53](https://gitlab.com/emd-dev/emd/-/merge_requests/53)
- Large number of pylint recommended fixes [271d7937](https://gitlab.com/emd-dev/emd/-/commit/271d793731fad64902f16323493ee06893002286)
- Indexing typo fixed in bin_by_phase [c5679432](https://gitlab.com/emd-dev/emd/-/commit/c5679432cfcd011965547144aaa936eee1405f62)
- Improve label alignments in plot_imfs [!54](https://gitlab.com/emd-dev/emd/-/merge_requests/56)

---

## 0.3.3

    pip install emd==0.3.3
Released 2021-02-04

### Added
- New function for computing summary stats from chains of cycles (from marcoFabus) [!46](https://gitlab.com/emd-dev/emd/-/merge_requests/46)

### Changed
- Major updates to tutorials [!40](https://gitlab.com/emd-dev/emd/-/merge_requests/40)
  - Binder notebooks added
  - New sifting tutorials added

### Fixed
- Replaced missing dependencies in setup.py [!42](https://gitlab.com/emd-dev/emd/-/merge_requests/42)

---

## 0.3.2

    pip install emd==0.3.2
Released 2020-11-29

### Added
- Add input array shape ensurance functions and start to use in sift & cycle submodules  [!26](https://gitlab.com/emd-dev/emd/-/merge_requests/26)
- Add more stopping criteria to sift module [!27](https://gitlab.com/emd-dev/emd/-/merge_requests/26)
  - Rilling et al and fixed iterations IMF stopping criteria
  - Energy threshold sift stopping criterion


### Changed
- Refactor some options extrema detection functions [!29](https://gitlab.com/emd-dev/emd/-/merge_requests/29)
- Sift throws an error if an IMF doesn't converge after a specified maximum number of iterations.
- Refactor mask generation in mask sift. Now specifies N masks of different phases and has options for parallel processing.
- SiftConfig yaml file also stores which type of sift the config is for [!35](https://gitlab.com/emd-dev/emd/-/merge_requests/35)
- 18% increase in testing coverage (to 75% total) [!30](https://gitlab.com/emd-dev/emd/-/merge_requests/30)

### Deprecated
- emd.spectra.frequency_stats renamed to emd.spectra.frequency_transform. Original func kept for now.

---

## 0.3.1

    pip install emd==0.3.1
Released 2020-09-06

### Added
- This changelog [!18](https://gitlab.com/emd-dev/emd/-/merge_requests/18)
- support.py submodule with some helper functions for checking installs and running tests [!20](https://gitlab.com/emd-dev/emd/-/merge_requests/20)
- envs subdir containing anaconda install environment config files [!21](https://gitlab.com/emd-dev/emd/-/merge_requests/21)
- Options for reading and writing sift configs to yaml format [!24](https://gitlab.com/emd-dev/emd/-/merge_requests/24)
- major update to webpage [!12](https://gitlab.com/emd-dev/emd/-/merge_requests/24)

utils_general.py
"""
General utilities
"""
import os
import logging
from datetime import datetime
from operator import attrgetter
from argparse import RawTextHelpFormatter, ArgumentDefaultsHelpFormatter
from pathlib import Path
import re
import numpy as np
from glob import glob
import pandas as pd


class SortedMenu(ArgumentDefaultsHelpFormatter):
    def add_arguments(self, actions):
        actions = sorted(actions, key=attrgetter('option_strings'))
        super(SortedMenu, self).add_arguments(actions)


class CustomFormatter(SortedMenu, RawTextHelpFormatter):
    pass


def create_parser(defaults):
    """
    Creates the parser object with the log arguments included.
    """
    try:
        import cli
        parser = cli.parser()
    except Exception:
        import argparse
        parser = argparse.ArgumentParser(formatter_class=CustomFormatter)

    parser.add_argument(
        "--logdir",
        type=str,
        default=defaults['logdir'],
        help="the log directory"
    )

    parser.add_argument(
        "--logname",
        type=str,
        default=None,
        help="the name of the logfile; specify is using the same logfile in append mode"
    )

    parser.add_argument(
        "--logfile",
        action="store_true",
        default=defaults['logfile'],
        help="create a log file"
    )
    parser.add_argument(
        "--no-logfile",
        dest="logfile",
        action="store_false",
        help="will not log on file"
    )
    parser.add_argument(
        "--logscreen",
        action="store_true",
        default=defaults['logscreen'],
        help="display the log on the screen"
    )
    parser.add_argument(
        "--no-logscreen",
        dest="logscreen",
        action="store_false",
        help="will not log on screen"
    )
    parser.add_argument(
        "--overwrite",
        action="store_true",
        default=defaults['overwrite'],
        help="overwrites the outputfile if it exists"
    )

    parser.add_argument(
        '--test',
        action='store_true',
        help='test run'
    )

    return parser


def makedir_if_needed(dir):
    if dir is not None:
        if not os.path.exists(dir):
            os.makedirs(dir)
    else:
        pass


def log_sysargs(logname, args, script=None, date=None,
                to_file=True, to_screen=False, filemode='a'):
    if date is None:
        date = datetime.now()
    if Path(logname).suffix != '.log':
        log_filename = '{0}_{1}.log'.format(logname,
                                            date.strftime('%Y%m%d-%H%M%S'))
    else:
        log_filename = logname

    if not Path(log_filename).parent.exists():
        makedir_if_needed(Path(log_filename).parent)

    if not os.path.exists(log_filename):
        filemode = 'w'

    handlers = []
    if to_file:
        fh = logging.FileHandler(log_filename, mode=filemode)  # new log everytime
        formatter = logging.Formatter(
                fmt='%(asctime)s %(message)s',
                datefmt='%m/%d/%Y %I:%M:%S %p',
            )
        fh.setFormatter(formatter)
        handlers.append(fh)

    if to_screen:
        sh = logging.StreamHandler()
        formatter = logging.Formatter(
                fmt='%(asctime)s %(message)s',
                datefmt='%m/%d/%Y %I:%M:%S %p',
            )
        sh.setFormatter(formatter)
        handlers.append(sh)

    if len(handlers) != 0:
        logging.basicConfig(
                # filename=log_filename,
                # filemode='w', # new log everytime
                # format='%(asctime)s %(message)s',
                # datefmt='%m/%d/%Y %I:%M:%S %p',
                level=logging.INFO,
                handlers=handlers,
            )

        if script is not None:
            info_text = '\n\n******SCRIPT PARAMETERS******\n\n'

            info_text += script + '\n'

            for k, v in vars(args).items():
                info_text += str(k) + '\t' + str(v) + '\n'
            info_text += '\n*****************************\n'

            logging.info(info_text)


def walk(path):
    """
    Get all files in subdirectories
    """
    for p in Path(path).iterdir():
        if p.is_dir():
            yield from walk(p)
            continue
        yield p.resolve()


def get_all_files(path):
    for p in walk(Path(path)):
        yield p


def glob_regex(pattern, path, sort=True):
    path = str(path)
    files = [f for f in os.listdir(path) if re.search(pattern, f)]
    if sort:
        files = sorted(files)
    return files


def get_files_in_dir(directory, test=False, seed=1234):
    directory = Path(directory)
    filelist = sorted(list(directory.iterdir()))
    if test:
        rng = np.random.default_rng(seed=seed)
        filelist = rng.choice(filelist, size=20, replace=False)
        filelist = filelist.tolist()

    return filelist


def get_pgids_in_dir(targetdir, ext='csv', suffix=None, as_frame=False):
    if suffix is None:
        search_term = str(targetdir / f'*.{ext}')
    else:
        search_term = str(targetdir / f'*{suffix}*.{ext}')

    pg_ids = [int(Path(i).stem.split('_')[0]) for i in glob(search_term)]

    srs = pd.Series(pg_ids).rename('pg_ids')

    if as_frame:
        srs = srs.to_frame()

    return srs


def rename_dict_keys(d, key_mapping):
    for orig, new in key_mapping.items():
        if orig in d.keys():
            d[new] = d.pop(orig)

    return d


def try_gz_suffix(filename, as_path=True):
    fname = Path(filename)
    if not fname.exists():
        fname = Path(f'{fname}.gz')

    try:
        assert fname.exists()
    except AssertionError:
        raise AssertionError(f'File does not exist: {fname}')

    if as_path:
        fname = Path(fname)
    else:
        fname = str(fname)

    return fname


def try_no_gz_suffix(filename, as_path=True):
    fname = Path(filename)
    if not fname.exists():
        fname = Path(str(fname).replace('.gz', ''))

    try:
        assert fname.exists()
    except AssertionError:
        raise AssertionError(f'File does not exist: {fname}')

    if as_path:
        fname = Path(fname)
    else:
        fname = str(fname)

    return fname


def get_seed_index(seed, srs, label=None):
    """
    Gets the row number of the seed in a seed file read as a pandas Series.
    """
    idx = srs[srs == seed].index
    if label is not None:
        label = 'Index'
    else:
        label = f'{label} index'
    try:
        assert len(idx) == 1
    except AssertionError:
        raise Exception(f'{label} invalid: {idx}')

    return idx


class NestedFunctions(object):
    """
    Functions with kwargs that are recast as functions of functions.
    Useful when used with pd.DataFrame.agg.
    """
    @staticmethod
    def percentile_fxn(percentile):
        def pctile(x):
            return np.percentile(x, percentile, axis=0)

        pctile.__name__ = f'pctile_{percentile}'

        return pctile


def split_arr_into_chunks(arr, chunksize):
    """
    Splits an array into chunks of length chunksize; the last chunk
    will be the remainder.
    """
    return np.split(arr, np.arange(chunksize, len(arr), chunksize))


class BootstrapGroupby(object):
    """Functions necessary for bootstrapping."""
    @classmethod
    def subsample_df(cls, df, grp_by='pg_id', n=40, seed=42):
        """
        Groups the dataframe by `grp_by` and subsamples, without replacement,
        `n` entries. In the case of rows corresponding to different seeds, this
        ensures that we have the same number of seeds for all `pg_ids`.
        """
        size_srs = df.groupby(grp_by).size()
        assert df.index.name == grp_by
        idx_pass = size_srs[size_srs >= n].index

        try:
            assert len(idx_pass) > 0
        except AssertionError:
            raise AssertionError(f'Not enough for subsampling: n={n}\n{df.head}')
        # try:
        #     assert size_srs.min() >= n
        # except AssertionError:
        #     msg = f'Minimum n in series is {size_srs.min()}. Replace n to proceed.'
        #     raise AssertionError(msg)

        subsample_df = df.loc[idx_pass].groupby(grp_by).apply(
            lambda grp: grp.sample(n=n, replace=False, random_state=seed)
            ).droplevel(1)

        return subsample_df

    @classmethod
    def generate_bootstrap_stats(cls, df, B, statistics,
                                 grp_by='pg_id',
                                 fillna=0, seed=None):
        """
        Can be applied to a dataframe or to a groupby object. We set fillna=0
        by default since we're working with NaN in variances of IMFs (an IMF
        that doesn't exist will have an energy of 0)

        Parameters
        ----------
        statistics: array-like
            list of functions that can be fed into pd.DataFrame.agg
        """
        df = df.fillna(fillna)  # copy
        sq = np.random.SeedSequence(seed)
        seedgen = sq.generate_state(B)
        stat_res = [None] * B
        for boot in range(B):
            rng = np.random.default_rng(seedgen[boot])
            boot_df = pd.DataFrame(
                df.values[rng.choice(df.shape[0], size=df.shape[0], replace=True)],
                columns=df.columns,
                index=df.index  # should all be the same, since this is for a particular pg_id
                )
            # takes too long
            # boot_df = df.sample(frac=1, replace=True, random_state=seedgen[boot])
            res = boot_df.agg(statistics)
            res.index.name = 'stat'
            res['boot_idx'] = boot
            stat_res[boot] = res
        stat_df = pd.concat(stat_res, axis=0).set_index('boot_idx', append=True).unstack(level=0)
        stat_df = stat_df.agg([
            np.mean,
            np.median,
            NestedFunctions.percentile_fxn(99.5),
            NestedFunctions.percentile_fxn(97.5),
            NestedFunctions.percentile_fxn(2.5),
            NestedFunctions.percentile_fxn(0.5),
        ])

        return stat_df


def get_seeds(seed_file=None, start_from_seed=None, end_at_seed=None,
              start_idx=None, end_idx=None, seeds=None):
    if seed_file is None and seeds is None:
        seeds = None

    elif seed_file is not None:
        seed_file = Path(seed_file)
        seeds = pd.read_csv(seed_file, header=None,
                            squeeze=True)
    else:
        seeds = seeds

    if seeds is not None:
        if start_from_seed is not None:
            assert start_idx is not None
            start_idx = get_seed_index(start_from_seed, seeds)

        if end_at_seed is not None:
            assert end_idx is not None
            end_idx = get_seed_index(end_at_seed, seeds)

        if start_idx is not None and end_idx is not None:
            seeds = seeds[start_idx:end_idx + 1]
        elif start_idx is not None and end_idx is None:
            seeds = seeds[start_idx]
        elif start_idx is None and end_idx is not None:
            seeds = seeds[:end_idx+1]
        else:
            pass

    return seeds


def nullable_string(val: str):
    if val.lower() in ['none', '']:
        return None
    else:
        return val


class sdict(dict):
    """
    Creates a dictionary where the keys are strings, an which accepts
    float and int keys, since it automatically converts them to strings.

    This ensures that the dictionary keys are not affected by how
    Python treats floats.
    """
    def __init__(self, *args, **kw):
        super(sdict,self).__init__(*args, **kw)
        # self.itemlist = super(odict,self).keys()

    def __getitem__(self, key):
        if not isinstance(key, str):
            key = str(key)
        return super().__getitem__(key)

    def __setitem__(self, key, value):
        if not isinstance(key, str):
            key = str(key)
        return super().__setitem__(key, value)


def get_window_skip_size_from_str(lbl):
    N_w = lbl.split('window=')[-1].split('_')[0]
    N_s = lbl.split('skip=')[-1].split('_')[0]

    try:
        N_w = int(N_w)
    except TypeError:
        pass

    try:
        N_s = int(N_s)
    except TypeError:
        pass

    return N_w, N_s
